lib/paypal_express/api.rb

```diff
@@ -58,18 +59,54 @@ def capture_payment(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb_
       def purchase_payment(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb_payment_method_id, amount, currency, properties, context)
         # Pass extra parameters for the gateway here
         options = {}
-
-        add_required_options(kb_payment_transaction_id, kb_payment_method_id, context, options)
-
-        properties        = merge_properties(properties, options)
-
-        # Can't use default implementation: the purchase signature is for one-off payments only
-        #super(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb_payment_method_id, amount, currency, properties, context)
-        gateway_call_proc = Proc.new do |gateway, linked_transaction, payment_source, amount_in_cents, options|
-          gateway.reference_transaction(amount_in_cents, options)
+        if find_value_from_properties(properties, 'from_hpp') == 'true'
+          options[:token] = find_value_from_properties(properties, 'token')
+
+          response = @response_model.create(:api_call                     => :build_form_descriptor,
+                                            :kb_account_id                => kb_account_id,
+                                            :kb_payment_id                => kb_payment_id,
+                                            :kb_payment_transaction_id    => kb_payment_transaction_id,
+                                            :transaction_type             => :PURCHASE,
+                                            :authorization                => options[:token],
+                                            :payment_processor_account_id => nil,
+                                            :kb_tenant_id                 => context.tenant_id,
+                                            :success                      => true,
+                                            :created_at                   => Time.now.utc,
+                                            :updated_at                   => Time.now.utc,
+                                            :message                      => { :payment_plugin_status => :PENDING }.to_json)
+          transaction          = response.to_transaction_info_plugin(nil)
+          transaction.status   = :PENDING
```

Comment:
Nit: I don't think you need to set it explicitly (since you set the `payment_plugin_status` above).